compiler-core/src/type_/printer.rs

```diff
@@ -92,18 +94,67 @@ pub struct TypeNames {
     /// value: `"something"`
     ///
     type_variables: HashMap<u64, EcoString>,
-    type_variable_names: HashSet<EcoString>,
+
+    /// Constructors which are imported in the current module in an
+    /// unqualified fashion.
+    ///
+    /// key:   (Defining module name, type name)
+    /// value: Alias name
+    ///
+    /// # Example 1
+    ///
+    /// ```gleam
+    /// import wibble.{Wobble}
+    /// ```
+    /// would result in
+    /// - key:   `("wibble", "Wobble")`
+    /// - value: `"Wobble"`
+    ///
+    /// # Example 2
+    ///
+    /// ```gleam
+    /// import wibble.{Wobble as Woo}
+    /// ```
+    /// would result in
+    /// - key:   `("wibble", "Wobble")`
+    /// - value: `"Woo"`
+    ///
+    local_constructors: HashMap<(EcoString, EcoString), EcoString>,
+
+    /// A map from local constructor names to the modules which they refer to.
+    /// This helps resolve cases like:
+    /// ```gleam
+    /// import wibble.{Wobble}
+    /// type Wibble { Wobble }
+    /// ```
+    /// Here, `Wobble` is shadowed, causing `Wobble` not to be valid syntax
+    /// for `wibble.Wobble`.
+    ///
+    /// Each key is the local name of the constructor, and the value is the module
+    /// for which the unqualified version is valid. In the above example,
+    /// it would result in
+    /// - key:   `"Wobble"`
+    /// - value: `"module"` (Whatever the current module is)
+    ///
+    /// But in this case:
+    /// ```gleam
+    /// import wibble.{Wobble as Wubble}
+    /// type Wibble { Wobble }
+    /// ```
+    /// No shadowing occurs, so this isn't needed.
+    ///
+    constructor_names: HashMap<EcoString, EcoString>,
 }
 
-impl TypeNames {
-    pub fn new(current_module: EcoString) -> Self {
+impl Names {
+    pub fn new() -> Self {
         Self {
-            uid: Default::default(),
-            current_module,
             local_types: Default::default(),
+            unshadowed_prelude_types: Default::default(),
             imported_modules: Default::default(),
             type_variables: Default::default(),
-            type_variable_names: Default::default(),
+            local_constructors: Default::default(),
```

Comment:
I think this is value constructors rather than type constructors so let's call them `local_value_constructors`